linalg/constrained.hpp

```diff
@@ -0,0 +1,272 @@
+    but with higher penalty the preconditioner is less effective. */
+class PenaltyConstrainedSolver : public IterativeSolver
+{
+public:
+   PenaltyConstrainedSolver(HypreParMatrix& A, SparseMatrix& B, double penalty_);
+
+   PenaltyConstrainedSolver(HypreParMatrix& A, HypreParMatrix& B, double penalty_);
+
+   ~PenaltyConstrainedSolver();
+
+   void Mult(const Vector& x, Vector& y) const;
+
+   void SetOperator(const Operator& op) { }
+
+private:
+   void Initialize(HypreParMatrix& A, HypreParMatrix& B);
+
+   double penalty;
+   Operator& constraintB;
+   HypreParMatrix * penalized_mat;
+   HypreBoomerAMG * prec;
+};
+
+#endif
+
+/** @brief Solve constrained system by solving original mixed sysetm;
+    see ConstrainedSolver.
+
+    Solves the saddle-point problem with a block-diagonal preconditioner, with
+    user-provided preconditioner in the top-left block and an identity matrix
+    in the bottom-right. */
+class SchurConstrainedSolver : public IterativeSolver
+{
+public:
+   SchurConstrainedSolver(BlockOperator& block_op_,
+                          Solver& primal_pc_);
+   virtual ~SchurConstrainedSolver();
+
+   void SetOperator(const Operator& op) { }
+
+   void Mult(const Vector& x, Vector& y) const;
+
+private:
+   BlockOperator& block_op;
+   Solver& primal_pc;
+   BlockDiagonalPreconditioner block_pc;
+   Solver * dual_pc;  // owned
+};
+
+
+/** @brief A class to solve the constrained system \f$ Ax = f \f$
+    subject to the constraint \f$ B x = r \f$ in an abstract setting.
+
+    Although this object may not use the below formulation, for
+    understanding some of its methods and notation you can think of
+    it as solving the saddle-point system
+
+     (  A   B^T  )  ( x )         (  f  )
+     (  B        )  ( lambda)  =  (  r  )
+
+     Not to be confused with ConstrainedOperator, which is totally
+     different. */
+class ConstrainedSolver : public Solver
+{
+public:
+   ConstrainedSolver(Operator& A, Operator& B);
+   ~ConstrainedSolver();
+
+   void SetOperator(const Operator& op) { }
+
+   /** @brief Setup Schur complement solver for constrained system.
+
+       @param prec Preconditioner for primal block.
+   */
+   void SetSchur(Solver& prec);
+
+   /** @brief Set the right-hand side r for the constraint B x = r
+
+       (r defaults to zero if you don't call this)
+   */
+   void SetDualRHS(const Vector& r);
+
+   /** @brief Set up the elimination solver.
+
+       The array secondary_dofs should contain as many entries as the rows
+       in the constraint matrix B; The block of B corresponding to these
+       columns will be inverted (or approximately inverted in a
+       preconditioner, depending on options) to eliminate the constraint.
+
+       @todo this can be done with only secondary_dofs given in interface
+   */
+   void SetElimination(Array<int>& primary_dofs,
+                       Array<int>& secondary_dofs);
+
+   /** @brief Set up a penalty solver. */
+   void SetPenalty(double penalty);
```

Comment:
I know jerome has brought this up, but it's possible that this penalty term could be different for every row (or even a sparse matrix itself). I think it would be fairly easy to expand on this interface for that option though.